### PR TITLE
MICROAPP-13067 SolarWinds

### PR DESCRIPTION
**Pull request by MichaelaVitusova**
new SolarWinds integration

### CODE REVIEW

**Comment by bogomil** [comment]
@leonidzhaldybin - can you give me an idea of when you can review this. I want to update the team on our standup.

**Review by leonidzhaldybin — CHANGES_REQUESTED** [review]

**Comment by leonidzhaldybin** [comment]
@bogomil Reviewed, sorry, was on PTO, and this one fell through the cracks.
Maybe asking someone from FE developers to have a look would be a good idea.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    let incidentComments = "";
+    let queryFields = "";
+    let exit = false;
+    let latestSync = new Date(latestSynchronizationTime);
+
+    if (latestSynchronizationTime === undefined)
+        queryFields = '?per_page=' + LIMIT + '&page=';
+    else {
+        queryFields = '?q=lastUpdateDate>=' + moment(latestSync).format('YYYY-MM-DD') + '&per_page=' + LIMIT + '&page=';
+    }
```
Comment (ondrejkralik) [inline bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js:70]:
```suggestion
    let queryFields = `?per_page=${LIMIT}`;

    if (!latestSynchronizationTime)
        queryFields = `${queryFields}&q=lastUpdateDate>=${moment(latestSynchronizationTime).format('YYYY-MM-DD')}`;
    }

// ....

    const resp = await client.fetch(`/incidents${queryFields}&page=${j}`);
    
```

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+        let incidentsEndpoint = ENDPOINT + queryFields + j;
+
+        let resp = await client.fetch(incidentsEndpoint);
+
+        if (resp.ok) {
+            var incidents = await resp.json();
+
+            if (incidents.length === 0) {
+                exit = true;
```
Comment (ondrejkralik) [inline bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js:92]:
```suggestion
                break;
```

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+            //Loop through each incident
+            for (i = 0; i < incidents.length; i++) {
+
+                let commentEndpoint = "";
+
+                if (latestSynchronizationTime === undefined)
+                    commentEndpoint = ENDPOINT + '/' + incidents[i].id + '/comments';
+                else {
+                    commentEndpoint = ENDPOINT + '/' + incidents[i].id + '/comments' + '?q=updated_at>=' + moment(latestSync).format('YYYY-MM-DD');
+                }
```
Comment (ondrejkralik) [inline bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js:104]:
```suggestion
                let commentEndpoint = `/incidents/${incidents[i].id}/comments`;

                if (!latestSynchronizationTime)
                    commentEndpoint = `${commentEndpoint}?q=updated_at>=${moment(latestSync).format('YYYY-MM-DD')}`;
                }
```

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+                var response = await client.fetch(commentEndpoint);
+
+                if (response.ok) {
+                    let comments = await response.json();
```
Comment (ondrejkralik) [inline bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js:109]: Can you give here an example of how `comments` response looks?
This looks very weird for me to convert to string and back.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    let str = value;
+    if (key === "body") {
+        str = str.replace(/<[^>]+>/gm, '');
+        str = str.replace('\n&nbsp;', '');
+    }
+    return str;
+}
+
+async function updateComments({dataStore, client, latestSynchronizationTime}) {
+    let incidentComments = "";
```
Comment (ondrejkralik) [inline bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js:61]: Why it is not array?

**Review by ondrejkralik — CHANGES_REQUESTED** [review]

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js (hunk 3 above)
Comment (ondrejkralik) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js:104]: You can prepare the whole thing before `for` loops
```js
const updatedAt = latestSynchronizationTime ? '' : `?q=updated_at>=${moment(latestSynchronizationTime).format('YYYY-MM-DD')}`;
```

and here you can do just
```js
var response = await client.fetch(`/incidents/${incidents[i].id}/comments${updatedAt}`);
```

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js (hunk 4 above)
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js:109]: It's been converted to string to apply a replace function.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js (hunk 5 above)
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js:61]: It is just another way to do.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/cd40c0ac-cc06-484d-9b52-e8aee2dfa759.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    let incidentComments = "";
+    let exit = false;
+    let latestSync = new Date(latestSynchronizationTime);
+    let updatedAt = '?q=updated_at>=' + moment(latestSync).format('YYYY-MM-DD');
+
+    let queryFields = '?q=lastUpdateDate>=' + moment(latestSync).format('YYYY-MM-DD') + '&per_page=' + LIMIT + '&page=';
+
+    if (latestSynchronizationTime === undefined) {
+        queryFields = '?per_page=' + LIMIT + '&page=';
+    }
```
Comment (leonidzhaldybin) [inline bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/cd40c0ac-cc06-484d-9b52-e8aee2dfa759.js:70]: Minor: if latestSynchronizationTime is undefined, the above code with calls to 'new Date' and `moment().format` is wasted. I'd place it to the `else` branch.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js (hunk 4 above)
Comment (leonidzhaldybin) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js]: Can't we iterate over comment objects and replace characters in their 'body' fields? Without converting the whole thing to string and back.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js (hunk 5 above)
Comment (ondrejkralik) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js]: So if there is no special reason then, please rewrite it to array.
Thank you.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/9ac77bce-a6c9-4177-ab55-23e364edb724.js [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+                        "primaryKey": true
+                    },
+                    {
+                        "name": "body",
+                        "type": "STRING",
+                        "length": 100
+                    },
+                    {
+                        "name": "created_at",
+                        "type": "STRING"
```
[the remark this answers is not in the document]
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/9ac77bce-a6c9-4177-ab55-23e364edb724.js]: Done.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/9ac77bce-a6c9-4177-ab55-23e364edb724.js [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+                        "name": "user_email",
+                        "type": "STRING"
+                    }
+                ]
+            }
+        ]
+    }
+});
+
+function replacer(key, value) {
```
[the remark this answers is not in the document]
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/9ac77bce-a6c9-4177-ab55-23e364edb724.js]: This function has been removed.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/9ac77bce-a6c9-4177-ab55-23e364edb724.js [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+    let latestSync = new Date(latestSynchronizationTime);
+
+    if (latestSynchronizationTime === undefined)
+        queryFields = '?per_page=' + LIMIT + '&page=';
+    else {
+        queryFields = '?q=lastUpdateDate>=' + latestSync.toISOString() + '&per_page=' + LIMIT + '&page=';
+    }
+
+    //Get page count
+    let urlEndpoint = ENDPOINT + queryFields + 0;
```
[the remark this answers is not in the document]
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/9ac77bce-a6c9-4177-ab55-23e364edb724.js]: Done.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/9ac77bce-a6c9-4177-ab55-23e364edb724.js [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+    let resp = await client.fetch(urlEndpoint);
+    if (resp.ok) {
+        PAGES = resp.headers["map"]["x-total-pages"];
```
[the remark this answers is not in the document]
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/9ac77bce-a6c9-4177-ab55-23e364edb724.js]: Done

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/9ac77bce-a6c9-4177-ab55-23e364edb724.js [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    //Loop through all pages
+    for (j = 0; j < PAGES; j++) {
+        let urlEndpoint = ENDPOINT + queryFields + j;
+
+        if (exit === true) {
+            break;
```
[the remark this answers is not in the document]
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/9ac77bce-a6c9-4177-ab55-23e364edb724.js]: Done.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/9ac77bce-a6c9-4177-ab55-23e364edb724.js [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+        }
+
+        urlEndpoint = parserURL(urlEndpoint, 'lastUpdateDate');
```
[the remark this answers is not in the document]
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/9ac77bce-a6c9-4177-ab55-23e364edb724.js]: Done.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/9ac77bce-a6c9-4177-ab55-23e364edb724.js [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+        let resp = await client.fetch(urlEndpoint);
+
+        if (resp.ok) {
+            var incidents = await resp.json();
+            var jsonIncidents = JSON.stringify(incidents, null, '\t');
+            var parsedIncidents = JSON.parse(jsonIncidents);
```
[the remark this answers is not in the document]
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/9ac77bce-a6c9-4177-ab55-23e364edb724.js]: That doesn't make sense. Removed.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/9ac77bce-a6c9-4177-ab55-23e364edb724.js [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+                }
+
+                commentEndpoint = parserURL(commentEndpoint, 'updated_at');
+
+                var response = await client.fetch(commentEndpoint);
+
+                console.log('i = ' + i +
+                    '\tticket id = ' + parsedIncidents[i].id +
+                    '\tticket number = ' + parsedIncidents[i].number +
+                    '\tstate = ' + parsedIncidents[i].state);
```
[the remark this answers is not in the document]
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/9ac77bce-a6c9-4177-ab55-23e364edb724.js]: Removed.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js (hunk 1 above)
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js]: Done.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js (hunk 2 above)
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js]: Done.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js (hunk 3 above)
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js]: Done.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js (hunk 4 above)
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js]: Function removed.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js (hunk 5 above)
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js]: Changed to array.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/cd40c0ac-cc06-484d-9b52-e8aee2dfa759.js [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+
+    //Get page count
+    let urlEndpoint = ENDPOINT + queryFields + 0;
+    let resp = await client.fetch(urlEndpoint);
+    const pages = resp.ok && resp.headers["map"]["x-total-pages"] ? resp.headers["map"]["x-total-pages"] : 1000;
+
+    //Loop through all pages
+    for (let j = 0; j < pages; j++) {
+
+        if (exit) {
```
[the remark this answers is not in the document]
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/cd40c0ac-cc06-484d-9b52-e8aee2dfa759.js:80]: Removed.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/cd40c0ac-cc06-484d-9b52-e8aee2dfa759.js [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+            break;
+        }
+
+        let incidentsEndpoint = ENDPOINT + queryFields + j;
+
+        let resp = await client.fetch(incidentsEndpoint);
```
[the remark this answers is not in the document]
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/cd40c0ac-cc06-484d-9b52-e8aee2dfa759.js:86]: Changed.

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/cd40c0ac-cc06-484d-9b52-e8aee2dfa759.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (cassiovanhelden) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/cd40c0ac-cc06-484d-9b52-e8aee2dfa759.js:80]: Removed.

File: bundles/http/Citrix/SolarWinds/baccf69b-821b-488c-b019-1ddb1d5297f0.js [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+            }
+        } else {
+            throw new Error(`Could not retrieve incidents (${incidentsPaginationResp.status}: ${incidentsPaginationResp.statusText})`);
+        }
+    }
+
+    if (incidentComments.length > 0) {
+        dataStore.save("updated_comments", incidentComments);
+    }
+}
```
Comment (ondrejkralik) [inline bundles/http/Citrix/SolarWinds/baccf69b-821b-488c-b019-1ddb1d5297f0.js:110]: This is how I would rewrite it.
I didn't test it, so if you want to use it as is, please check that everything is ok.
I presume that I was right about `updatedAt ` and `lastUpdateDate` so I changed the logic.
The next thing which I changed was to check `response.ok` and throw error immediately after call. IMHO it is more readable than if/else.

```js
async function updateComments({dataStore, client, latestSynchronizationTime}) {
    const incidentComments = [];
    const latestSync = moment(latestSync).format('YYYY-MM-DD');
    const updatedAt = latestSynchronizationTime ? `?q=updated_at>=${latestSync}` : '';
    const lastUpdateDate = latestSynchronizationTime ? `?q=lastUpdateDate>=${latestSync}` : '';
    
    const incidentsResp = await client.fetch(`/incidents?per_page=${LIMIT}${lastUpdateDate}`);

    if (!incidentsResp.ok) {
        throw new Error(`Could not retrieve incidents count (${incidentsResp.status}: ${incidentsResp.statusText})`);
   }

    const pagesCount = incidentsResp.headers['map']['x-total-pages'] ?? 1000;

    //Loop through all pages
    for (let j = 0; j < pagesCount; j++) {
        const incidentsPaginationResp = await client.fetch(`/incidents?per_page=${LIMIT}${lastUpdateDate}&page=${j}`);

        if (!incidentsPaginationResp.ok) {
            throw new Error(`Could not retrieve incidents (${incidentsPaginationResp.status}: ${incidentsPaginationResp.statusText})`);
        }

        const incidents = await incidentsPaginationResp.json();
        if (incidents.length === 0) {
            break;
        }

        //Loop through each incident
        for (const incident of incidents) {
            const commentsResponse = await client.fetch(`/incidents/${incident.id}/comments${updatedAt}`);

            if (!commentsResponse.ok) {
                throw new Error(`Could not retrieve comments (${commentsResponse.status}: ${commentsResponse.statusText})`);
            }

            let commentsJson = await commentsResponse.json();

            if (commentsJson) {
                commentsJson.forEach(comment => {
                    incidentComments.push({
                        id: comment.id,
                        body: comment.body.replace(/<\/?([a-z][a-z0-9]*)\b[^>]*>/gi, ''),
                        created_at: new Date(comment.created_at),
                        commenter_id: comment.commenter_id,
                        user_name: comment.user.name,
                        user_email: comment.user.email,
                    });
                });
            }
        }
    }

    if (incidentComments.length > 0) {
        dataStore.save('updated_comments', incidentComments);
    }
}
```

File: bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js (hunk 3 above)
Comment (ondrejkralik) [reply bundles/http/Citrix/5c20233e-ead7-4385-a3f9-c703f194514b/5b0d4f89-2cfb-499b-8f63-f70cbed8fa13.js]: I see.
The condition which I suggested here has the opposite meaning here and you use it in the latest change.
Sorry for confusing you.

**Comment by bogomil** [comment]
@ondrejkralik  - there is a new PR fixing everything you asked for :)

**Review by ondrejkralik — APPROVED** [review]

**Review by petrdytrychcitrix — APPROVED** [review]

**Review by leonidzhaldybin — APPROVED** [review]

**Pull request merged**